config/initializers/redis_hacks.rb

```diff
@@ -4,34 +4,36 @@
 # See https://github.com/redis/redis-rb/pull/856
 # We can remove this once we can upgrade the gem to >= v4.1.2
```

Comment:
I should have removed this comment in https://github.com/3scale/porta/pull/1700. You may want to do it in the PR maybe.

It's a shame that now that we can upgrade the gem to a version where sentinel passwords _are_ supported, we still need to keep this hack only because their implementation doesn't properly handle DNS failures such as the ones reported in https://issues.redhat.com/browse/THREESCALE-4495.

Reply:
Ah it is another issue then, I kinda remember some conversation about it, I did not make the link.
We can make a PR to the upstream, Porta can wait for it before merging this PR :)
Wdyt @guicassolato?

Reply:
@guicassolato They fixed it in this updated version.

I made a simple test.

```ruby
# frozen_string_literal: true
require 'bundler'
Bundler.require

SENTINELS = [{ host: "nonexistent", port: 26380 },
  { host: "unreachable", port: 26381 }]

redis = Redis.new(url: "redis://unreachablemaster", sentinels: SENTINELS, role: :master)
```

With redis `4.1.3` it raise a `Redis::CannotConnectError`

```shell
hery@Tiburon ~/Devel/Ruby/redis-test $ BUNDLE_GEMFILE=gemfiles/4.1.3/Gemfile ruby test.rb                       [2.6]
Traceback (most recent call last):
	17: from test.rb:9:in `<main>'
	16: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis.rb:152:in `ping'
	15: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis.rb:52:in `synchronize'
	14: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
	13: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis.rb:52:in `block in synchronize'
	12: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis.rb:153:in `block in ping'
	11: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:125:in `call'
	10: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:230:in `process'
	 9: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:319:in `logging'
	 8: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:231:in `block in process'
	 7: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:381:in `ensure_connected'
	 6: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:105:in `connect'
	 5: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:306:in `with_reconnect'
	 4: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:106:in `block in connect'
	 3: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:344:in `establish_connection'
	 2: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:565:in `resolve'
	 1: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:602:in `resolve_master'
/Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-4.1.3/lib/redis/client.rb:598:in `sentinel_detect': No sentinels available. (Redis::CannotConnectError)
```

With redis `3.3.5` (our current master) it raises `SocketError`

```shell
hery@Tiburon ~/Devel/Ruby/redis-test $ BUNDLE_GEMFILE=gemfiles/3.3.5/Gemfile ruby test.rb                       [2.6]
Traceback (most recent call last):
	32: from test.rb:9:in `<main>'
	31: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis.rb:148:in `ping'
	30: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis.rb:58:in `synchronize'
	29: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
	28: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis.rb:58:in `block in synchronize'
	27: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis.rb:149:in `block in ping'
	26: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:120:in `call'
	25: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:220:in `process'
	24: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:306:in `logging'
	23: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:221:in `block in process'
	22: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:364:in `ensure_connected'
	21: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:100:in `connect'
	20: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:293:in `with_reconnect'
	19: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:101:in `block in connect'
	18: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:331:in `establish_connection'
	17: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:535:in `resolve'
	16: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:571:in `resolve_master'
	15: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:546:in `sentinel_detect'
	14: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:546:in `each'
	13: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:554:in `block in sentinel_detect'
	12: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:572:in `block in resolve_master'
	11: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:120:in `call'
	10: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:220:in `process'
	 9: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:306:in `logging'
	 8: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:221:in `block in process'
	 7: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:364:in `ensure_connected'
	 6: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:100:in `connect'
	 5: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:293:in `with_reconnect'
	 4: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:101:in `block in connect'
	 3: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/client.rb:336:in `establish_connection'
	 2: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/connection/ruby.rb:313:in `connect'
	 1: from /Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/connection/ruby.rb:229:in `connect'
/Users/hery/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/redis-3.3.5/lib/redis/connection/ruby.rb:229:in `getaddrinfo': getaddrinfo: nodename nor servname provided, or not known (SocketError)
```

Reply:
We we checked together this morning, the problem is that https://github.com/redis/hiredis-rb/blob/5f0eb41b8fac37fee94cfbe3af482b904c82e48e/lib/hiredis/ruby/connection.rb#L192 won't raise a `SockerError`, but a `RuntimeError` instead.

From their README:
> When the connection was closed by the server, an error of the type Hiredis::Connection::EOFError will be raised. For all I/O related errors, the Ruby built-in Errno::XYZ errors will be raised. All other errors (such as a protocol error) result in a RuntimeError.

Reply:
We can upgrade redis-rb, but we'd also need to fix hiredis in to make it raise the proper error in this case.

Reply:
The issue is on hiredis gem https://github.com/redis/hiredis-rb#standalone-connection

> When the connection was closed by the server, an error of the type Hiredis::Connection::EOFError will be raised. For all I/O related errors, the Ruby built-in Errno::XYZ errors will be raised. All other errors (such as a protocol error) result in a RuntimeError.